src/compiler/visitorPublic.ts

```diff
@@ -22,35 +22,61 @@ import {
+ * - If the visitor returns undefined, then the output is undefined.
+ * - If the output node is not undefined, then it will satisfy the test function.
+ * - In order to obtain a return type that is more specific than `Node`, a test
+ *   function _must_ be provided, and that function must be a type predicate.
+ *
  * @param node The Node to visit.
  * @param visitor The callback used to visit the Node.
  * @param test A callback to execute to verify the Node is valid.
  * @param lift An optional callback to execute to lift a NodeArray into a valid Node.
  */
-export function visitNode<T extends Node>(node: T, visitor: Visitor | undefined, test?: (node: Node) => boolean, lift?: (node: readonly Node[]) => T): T;
-
+export function visitNode<TIn extends Node | undefined, TVisited extends Node | undefined, TOut extends Node>(
+    node: TIn,
+    visitor: Visitor<NonNullable<TIn>, TVisited> | undefined,
+    test: (node: Node) => node is TOut,
+    lift?: (node: readonly Node[]) => Node,
+): TOut | (TIn & undefined) | (TVisited & undefined);
 /**
  * Visits a Node using the supplied visitor, possibly returning a new Node in its place.
  *
+ * - If the input node is undefined, then the output is undefined.
+ * - If the visitor returns undefined, then the output is undefined.
+ * - If the output node is not undefined, then it will satisfy the test function.
+ * - In order to obtain a return type that is more specific than `Node`, a test
+ *   function _must_ be provided, and that function must be a type predicate.
+ *
  * @param node The Node to visit.
  * @param visitor The callback used to visit the Node.
  * @param test A callback to execute to verify the Node is valid.
  * @param lift An optional callback to execute to lift a NodeArray into a valid Node.
  */
-export function visitNode<T extends Node>(node: T | undefined, visitor: Visitor | undefined, test?: (node: Node) => boolean, lift?: (node: readonly Node[]) => T): T | undefined;
-
-export function visitNode<T extends Node>(node: T | undefined, visitor: Visitor | undefined, test?: (node: Node) => boolean, lift?: (node: readonly Node[]) => T): T | undefined {
-    if (node === undefined || visitor === undefined) {
+export function visitNode<TIn extends Node | undefined, TVisited extends Node | undefined>(
+    node: TIn,
+    visitor: Visitor<NonNullable<TIn>, TVisited> | undefined,
+    test?: (node: Node) => boolean,
+    lift?: (node: readonly Node[]) => Node,
+): Node | (TIn & undefined) | (TVisited & undefined);
+export function visitNode(
+    node: Node,
+    visitor: Visitor | undefined,
+    test?: (node: Node) => boolean,
+    lift?: (node: readonly Node[]) => Node,
+): Node | undefined {
+    if (node === undefined) {
+            // If the input type is undefined, then the output type can be undefined.
         return node;
     }
 
-    const visited = visitor(node);
-    if (visited === node) {
-        return node;
-    }
+    const visited = visitor ? visitor(node) : node;

```

Comment:
The only time `visitor` should be `undefined` in our codebase is when `visitNode` is called for a token in `visitEachChild`. And that's only really supposed to be for position tracking when making text changes in services. Perhaps we should make `visitor` non-optional and just conditionally call `visitNode` for tokens below.

Reply:
Replacing the uses below with `tokenVisitor ? visitNode(node.someToken, tokenVisitor, isSomeToken) : node.someToken` works. The only other place to fix is in `destructuring.ts`'s flattening code, which can be changed in a similar way. When making those changes, all tests pass.

My only qualms about this are the same as with `test`; there is likely existing code out there that  should continue to work, which means that we probably should leave the checks in here and maintain behavior if possible. But if we are willing to do that, I'm not so sure that we really need to remove them from the API like that either. On the fence...

Reply:
I have the change for this staged, if we want to make the visitor required, but my question about whether or not we still want to check if the visitor is defined or not is still of interest.

Reply:
I did a [SourceGraph search](https://sourcegraph.com/search?q=context:global+ts%5C.visitNode%5C%28%5Cw%2B%5C%29+lang:TypeScript+-file:.*%5C.d%5C.ts&patternType=regexp&sm=1&groupBy=repo) and see no users who are passing in undefined. I'll make it required.